var/spack/repos/builtin/packages/openfoam/package.py

```diff
@@ -0,0 +1,358 @@
+##############################################################################
+# Copyright (c) 2017 OpenCFD Ltd.
+#
+# This file is released as part of Spack.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for the LLNL notice and the LGPL.
+#
+# License
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+from spack.environment import *
+
+import multiprocessing
+import glob
+import os
+import re
+import shutil
+
+
+# Notes.
+# - Resolution of flex, zlib needs more attention
+# - Paraview/VTK build - needs adjustments in ~/.spack/packages.yaml
+class Openfoam(Package):
+    """OpenFOAM is a GPL-opensource C++ CFD-toolbox.
+       OpenCFD Ltd has been developing and releasing OpenFOAM since its debut in 2004."""
+
+    homepage = "http://www.openfoam.com/"
+    base_url = "https://sourceforge.net/projects/openfoamplus/files"
+
+    version('1612', 'ca02c491369150ab127cbb88ec60fbdf',
+        url = base_url+'/v1612+/OpenFOAM-v1612+.tgz')
+
+    variant('scotch', default=True,
+        description='With scotch for decomposition')
+    variant('ptscotch', default=True,
+        description='With pt-scotch for decomposition')
+    variant('metis', default=True,
+        description='With metis for decomposition')
+    ## Not yet working properly
+    ##    variant('parmgridgen', default=True,  ## Not yet working properly
+    ##        description='With parmgridgen support')
+    variant(
+        'source', default=True,
+        description='Install library/application sources and tutorials')
+
+    variant('paraview', default=True,
+        description='Build paraview plugins and runtime post-processing')
+
+    supported_compilers = {'clang': 'Clang', 'gcc': 'Gcc', 'intel': 'Icc'}
+
+    depends_on('mpi')
+    depends_on('flex')
+    depends_on('zlib')
+    depends_on('fftw')
+    depends_on('boost')
+    depends_on('cgal')
+    depends_on('cmake', type='build')
+
+    depends_on('scotch ~ metis', when='~ptscotch+scotch')
+    depends_on('scotch ~ metis + mpi', when='+ptscotch')
```

Comment:
Sure, I just need to understand the syntax more (I copied this from elsewhere). I guess this means that I need scotch without the metis variant, but only when scotch has been selected and ptscotch deselected?  I can't really figure out why that is useful or needed.
Maybe it's supposed to require "scotch~metis" when the +metis variant has been selected?

Assuming that I get that sorted, what happens when I wish to add another constraint.
For example, if I'm building with +int64, I would like to require scotch also have +int64, but still retain the previous (faulty?) logic for ~ptscotch.


  - depend on scotch+int64